L1Trigger/Phase2L1ParticleFlow/interface/taus/L1HPSPFTauEmulator.h

```diff
@@ -0,0 +1,426 @@
+      } else {
+        if (is_charged(part) && int_abs(deltaZ) > dzCut) {
+          iso_pt += part.hwPt;
+          inCone = false;
+        }
+      }
+    }
+    return inCone;
+  }
+
+  inline Tau makeHPSTauHW(const std::vector<Particle>& parts,
+                          const Particle seed,
+                          const pt_t total_pt /*, Config config*/) {
+    using namespace l1HPSPFTauEmu;
+
+    ap_uint<20> scone2 = setSConeSize2(total_pt);
+
+    pt_t isocone_pt = 0;
+
+    pt_t sum_pt = 0;
+
+    ap_fixed<22, 20> sum_eta = 0;
+    ap_fixed<22, 20> sum_phi = 0;
+
+    pt_t tau_pt = 0;
+    etaphi_t tau_eta = 0;
+    etaphi_t tau_phi = 0;
+
+    pt_t chargedIsoPileup = 0;
+    std::vector<Particle> signalParts;
+    std::vector<Particle> outsideParts;
+
+    int trct = 0;
+    bool leadCand = false;
+    bool leadSet = false;
+    Particle lead;
+    for (std::vector<int>::size_type i = 0; i != parts.size(); i++) {
+      bool isSignal = inSignalCone(parts.at(i), seed, trct, scone2, isocone_pt, leadCand);
+      if (isSignal) {
+        signalParts.push_back(parts.at(i));
+        if (parts[i].pID == 0) {
+          trct++;
+        }
+        if (leadCand) {
+          if (leadSet == false) {
+            lead = parts[i];
+            leadCand = false;
+            leadSet = true;
+          } else {
+            if (parts[i].hwPt > lead.hwPt) {
+              lead = parts[i];
+              leadCand = false;
+            } else {
+              leadCand = false;
+            }
+          }
+        }
+      } else {
+        outsideParts.push_back(parts.at(i));
+      }
+    }
+
+    for (std::vector<int>::size_type i = 0; i != signalParts.size(); i++) {
+      Particle sigP = signalParts.at(i);
+      if (is_charged(sigP) || (sigP.pID == 3)) {
+        sum_pt += sigP.hwPt;
+        sum_eta += sigP.hwPt * sigP.hwEta;
+        sum_phi += sigP.hwPt * sigP.hwPhi;
+      }
+    }
+
+    pt_t div_pt = 1;
+    if (sum_pt == 0) {
+      div_pt = 1;
+    } else {
+      div_pt = sum_pt;
+    }
+
+    tau_pt = sum_pt;
+    tau_eta = sum_eta / div_pt;
+    tau_phi = sum_phi / div_pt;
+
+    if (tau_pt > l1ct::Scales::makePtFromFloat(20.) && int_abs(tau_eta) < etaCutoff && leadSet == true &&
+        isocone_pt < tau_pt) {
+      Tau tau;
+      tau.hwPt = tau_pt;
+      tau.hwEta = tau_eta;
+      tau.hwPhi = tau_phi;
+      return tau;
+    } else {
+      Tau tau;
+      tau.hwPt = 0.;
+      tau.hwEta = 0.;
+      tau.hwPhi = 0.;
+      return tau;
+    }
+  }
+
+  inline std::vector<Tau> emulateEvent(std::vector<Particle>& parts, std::vector<Particle>& jets, bool jEnable) {
+    using namespace l1HPSPFTauEmu;
+
+    std::vector<Particle> parts_copy;
+    parts_copy.resize(parts.size());
+    std::transform(parts.begin(), parts.end(), parts_copy.begin(), [](const Particle& part) { return part; });
+    //sorting by pt
+    std::sort(parts_copy.begin(), parts_copy.end(), [](Particle i, Particle j) { return (i.hwPt > j.hwPt); });
```

Comment:
Please optimize a bit here:
std::sort(parts_copy.begin(), parts_copy.end(), [](const Particle& i, const Particle& j) { return (i.hwPt > j.hwPt); });

Similarly in other std::sort(...)